### PR TITLE
RIP and EIP relative display

### PR DESCRIPTION
**Pull request by mumbel**
The display should print the register and immediate offset
instead of the relative value.

### CODE REVIEW

File: Ghidra/Processors/x86/data/languages/ia.sinc
Comment (ghidra1) [inline Ghidra/Processors/x86/data/languages/ia.sinc:686]: This change will prevent our operand markup from properly replacing the constant with the reference label and will force the use of an extended markup using ```=>```.  In all of our language implementations we generally prefer producing a computed constant value for PC relative addresses.  I realize the original instruction form is muddled in doing this.  I will defer to @emteere for his opinion on the matter.

Comment (emteere) [reply Ghidra/Processors/x86/data/languages/ia.sinc:686]: The result of this change will be:
[ RIP + simm ]=>LABEL
Instead of:
[ LABEL ]

**Comment by jobermayr** [comment]
The patch doesn't work on current master:
image

Without:
image

File: Ghidra/Processors/x86/data/languages/ia.sinc
Comment (ghidra1) [reply Ghidra/Processors/x86/data/languages/ia.sinc:686]: The change as it stands will prevent a DEFAULT reference and will need to be computed via analysis. 

While it may be possibe to alter the representation and live with the `=>` markup, the exported constant should remain as it was so that the instruction can emit the absolute Scalar.

**Comment by ghidra1** [comment]
> The patch doesn't work on current master
It certainly appears to be having the desired affect that @mumbel was after.  What about this do you feel is wrong?

**Comment by jobermayr** [comment]
> [...] What about this do you feel is wrong?
In decompiler window:
`- PcInitializeAdapterDriver(param_1,param_2,&LAB_00011008)`
`+ (*(code *)0xfd05634c38ec8348(param_1,param_2,&LAB_00011008)`

File: Ghidra/Processors/x86/data/languages/ia.sinc
Comment (ghidra1) [reply Ghidra/Processors/x86/data/languages/ia.sinc:686]: Anything which can be computed as a constant value or address at parse time should be exported as such.  It is conceivable that analysis may place the reference on the incorrect operand if this is not done - although I suppose it may not matter much.  In either case, since a constant was produced and not an Address, it will require an analyzer to laydown the reference - although possibly a different analzer for the two cases.  As a constant value, there may be some speculating if the constant refers to a memory location or not.  This would depend on evidence of de-referencing or use of an associated pointer data type.  

I suspect the net affect is the same other than the use of an extended `=>` reference markup and possibly on a different operand.

Comment (ghidra1) [reply Ghidra/Processors/x86/data/languages/ia.sinc:686]: Still waiting on @emteere for the pass/fail decision.

**Comment by ghidra1** [comment]
@mumbel The pcode is invalid:
```
{ local riprel:4 = inst_next + simm32; export *[const]:4 riprel; }
```
The local variable `riprel` is a unqiue varnode and can't be export as a constant.  The correct subconstructor should look like this (applies to other changes as well):
```
addr32: [RIP + simm32]						is bit64=1 & mod=0 & r_m=4; index=4 & base=5; simm32 & RIP [riprel=inst_next+simm32; ] { export *[const]:4 riprel; }
```

**Review by ghidra1 — CHANGES_REQUESTED** [review]
Pcode has invalid constant export which must be pre-computed in `[  ]`s as it was originally

**Comment by ghidra1** [comment]
Although my suggested fix above is not quite right, I have put this PR in for merge and have made the neccessary correction for the constant export issue.

**Review by ghidra1 — COMMENTED** [review]
Additional changes will be made prior to merge.

File: Ghidra/Processors/x86/data/languages/ia.sinc
Comment (ghidra1) [reply Ghidra/Processors/x86/data/languages/ia.sinc:686]: > Doesn't this just show errors in analysis not the other way around?

See issues raised by #4892

Comment (ghidra1) [reply Ghidra/Processors/x86/data/languages/ia.sinc:686]: @mumbel could you please provide you arguments for supporting this PR within the PR #4892 discussion which seeks to revert this change.  Unless there is a compelling reason we are likely to revert the changes made by your PR.  Thanks.

Comment (mumbel) [reply Ghidra/Processors/x86/data/languages/ia.sinc:686]: My catalyst for doing this was 2 1/2 years ago and I do not recall the situation.  My interpretation of the manual is that the register should appear and it's relative offsets.  The other thread seems to completely support that in both compiling (assembling) and in disassembly.  Unless I've missed it, no one has given a counter argument to this change in terms of correctness.

The current issue appears to be outside of sleight/pcode (right?), so I'm not sure what any comments I have on this instruction change matter.

It's your project, if you need to revert, revert

Comment (ghidra1) [reply Ghidra/Processors/x86/data/languages/ia.sinc:686]: The biggest issue stems from the impact to the patching assembler which has no real way at present to contend with RIP-relative operands when its is simplest to specify an address/symbol.  There is no obvious way for the assembler to identify the relative offset bias from the language spec.  The pcode is the same in both cases.

**Pull request closed without merging**
